packages/next/build/webpack/plugins/next-trace-entrypoints-plugin.ts

```diff
@@ -0,0 +1,239 @@
+}
+
+export class TraceEntryPointsPlugin implements webpack.Plugin {
+  private appDir: string
+  private entryTraces: Map<string, string[]>
+
+  constructor({ appDir }: { appDir: string }) {
+    this.appDir = appDir
+    this.entryTraces = new Map()
+  }
+
+  // Here we output all traced assets and webpack chunks to a
+  // ${page}.nft.json file
+  createTraceAssets(compilation: any, assets: any) {
+    const outputPath = compilation.outputOptions.path
+
+    for (const entrypoint of compilation.entrypoints.values()) {
+      const entryFiles = new Set<string>()
+
+      for (const chunk of entrypoint
+        .getEntrypointChunk()
+        .getAllReferencedChunks()) {
+        for (const file of chunk.files) {
+          entryFiles.add(nodePath.join(outputPath, file))
+        }
+        for (const file of chunk.auxiliaryFiles) {
+          entryFiles.add(nodePath.join(outputPath, file))
+        }
+      }
+      // don't include the entry itself in the trace
+      entryFiles.delete(
+        nodePath.join(
+          outputPath,
+          `${isWebpack5 ? '../' : ''}${entrypoint.name}.js`
+        )
+      )
+
+      assets[
+        `${isWebpack5 ? '../' : ''}${entrypoint.name}.nft.json`
+      ] = new sources.RawSource(
+        JSON.stringify({
+          version: TRACE_OUTPUT_VERSION,
+          files: [
+            ...entryFiles,
+            ...this.entryTraces.get(entrypoint.name)!,
+          ].map((file) => file.replace(/\\/g, '/')),
+        })
+      )
+    }
+  }
+
+  apply(compiler: webpack.Compiler) {
+    if (isWebpack5) {
+      compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
+        // @ts-ignore TODO: Remove ignore when webpack 5 is stable
+        compilation.hooks.processAssets.tap(
+          {
+            name: PLUGIN_NAME,
+            // @ts-ignore TODO: Remove ignore when webpack 5 is stable
+            stage: webpack.Compilation.PROCESS_ASSETS_STAGE_ADDITIONS,
+          },
+          (assets: any) => {
+            this.createTraceAssets(compilation, assets)
+          }
+        )
+      })
+    } else {
+      compiler.hooks.emit.tap(PLUGIN_NAME, (compilation: any) => {
+        this.createTraceAssets(compilation, compilation.assets)
+      })
+    }
+
+    compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
+      compilation.hooks.finishModules.tapAsync(
+        PLUGIN_NAME,
+        async (_stats: any, callback: any) => {
+          // we create entry -> module maps so that we can
+          // look them up faster instead of having to iterate
+          // over the compilation modules list
+          const entryNameMap = new Map<string, string>()
+          const entryModMap = new Map<string, any>()
+
+          try {
+            const depModMap = new Map<string, any>()
+
+            compilation.entries.forEach((entry) => {
+              const name = entry.name || entry.options?.name
+
+              if (name?.startsWith('pages/') && entry.dependencies[0]) {
```

Comment:
Here you should iterate over all `entry.dependencies` to handle cases like `entry: { "pages/a": ["./pages/a.js", "./something-else.js"] }`